File: examples/go/match.go
// A runnable Go example: index a history and match the current state.
//
//	cargo build --release -p wickra-shazam-c
//	# stage the library under bindings/go/lib/<goos>_<goarch>/ (CI does this)
//	cd examples/go && go run .
package main

import (
	"fmt"

	wickra "github.com/wickra-lib/wickra-shazam-go"
)

const spec = `{"features":[{"kind":"price","field":"close"}],` +
	`"window":1,"metric":"euclid"}`

const indexCmd = `{"cmd":"index","history":[` +
	`{"time":1,"open":100,"high":100,"low":100,"close":100,"volume":1},` +
	`{"time":2,"open":101,"high":101,"low":101,"close":101,"volume":1},` +
	`{"time":3,"open":102,"high":102,"low":102,"close":102,"volume":1}]}`

const matchCmd = `{"cmd":"match","current":[` +
	`{"time":4,"open":102,"high":102,"low":102,"close":102,"volume":1}],"k":2}`

func main() {
	shazam, err := wickra.New(spec)
	if err != nil {
		panic(err)
	}
	defer shazam.Close()

	if _, err := shazam.Command(indexCmd); err != nil {
		panic(err)
	}
	report, err := shazam.Command(matchCmd)
	if err != nil {
		panic(err)
	}

	fmt.Println("wickra-shazam", wickra.Version())
	fmt.Println(report)
}

File: examples/java/Match.java
// A runnable Java example: index a history and match the current state.
//
//   cargo build -p wickra-shazam-c
//   mvn -f bindings/java/pom.xml -q package -DskipTests
//   javac -cp bindings/java/target/classes examples/java/Match.java -d examples/java/out
//   java --enable-native-access=ALL-UNNAMED \
//        -Dnative.lib.dir=target/debug \
//        -cp "bindings/java/target/classes;examples/java/out" Match
import org.wickra.shazam.Shazam;

public final class Match {
    private static final String SPEC =
            "{\"features\":[{\"kind\":\"price\",\"field\":\"close\"}],"
                    + "\"window\":1,\"metric\":\"euclid\"}";

    private static String candle(int time, int close) {
        return "{\"time\":" + time + ",\"open\":" + close + ",\"high\":" + close
                + ",\"low\":" + close + ",\"close\":" + close + ",\"volume\":1}";
    }

    public static void main(String[] args) {
        try (Shazam shazam = new Shazam(SPEC)) {
            String history = "[" + candle(1, 100) + "," + candle(2, 101) + "," + candle(3, 102) + "]";
            shazam.command("{\"cmd\":\"index\",\"history\":" + history + "}");
            String response =
                    shazam.command("{\"cmd\":\"match\",\"current\":[" + candle(4, 102) + "],\"k\":2}");
            System.out.println("wickra-shazam " + Shazam.version());
            System.out.println(response);
        }
    }
}
